### PR TITLE
feat: Creating Gardener Cluster CR as part of cluster provisioning

### PR DESCRIPTION
**Pull request by koala7659**
**Description**

Changes proposed in this pull request:

- Creating Gardener Cluster CR as next stage of Provisioning
- Validation of labels on provided Runtime CR: 
Required labels: 

```
- kyma-project.io/instance-id
- kyma-project.io/runtime-id
- kyma-project.io/region
- operator.kyma-project.io/kyma-name
- kyma-project.io/broker-plan-id
- kyma-project.io/broker-plan-name
- kyma-project.io/global-account-id
- kyma-project.io/subaccount-id
```

- State machine optimisation

**Testing:**

- Unit tests for State machine are not fully implemented
- Integration tests are in progress

**Related issue(s)**
https://github.com/kyma-project/infrastructure-manager/issues/282

### CODE REVIEW

File: internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+				imv1.LabelKymaGlobalSubaccountID: runtime.Labels[imv1.LabelKymaGlobalSubaccountID], // BTW most likely this value will be missing
+				imv1.LabelKymaName:               runtime.Labels[imv1.LabelKymaName],
+
+				// values from Runtime CR fields
+				imv1.LabelKymaPlatformRegion: runtime.Spec.Shoot.PlatformRegion,
+				imv1.LabelKymaRegion:         runtime.Spec.Shoot.Region,
+				imv1.LabelKymaShootName:      shoot.Name,
+
+				// hardcoded values
+				imv1.LabelKymaManagedBy: "lifecycle-manager",
```
Comment (Disper) [inline internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig.go:87]: KEB sets that currently to `lifecycle-manager which is wrong. KIM is managing this resource. It should be set to `infrastructure-manager`

Comment (Disper) [inline internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig.go:87]:
```suggestion
				imv1.LabelKymaManagedBy: "infrastructure-manager",
```
KEB sets that currently to `lifecycle-manager` which is wrong. KIM is managing this resource. It should be set to `infrastructure-manager`

Comment (koala7659) [reply internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig.go]: Fixed

File: internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig_test.go [hunk 2]
```diff
@@ -1 +1,164 @@
...
+			testCtx,
+			must(newFakeFSM, withTestFinalizer),
+			&systemState{instance: testRtWithDeletionTimestampAndFinalizer, shoot: &testShoot},
+			testOpts{
+				MatchExpectedErr: BeNil(),
+				MatchNextFnState: haveName("sFnDeleteShoot"),
+			},
+		),
+		Entry(
+			"should return sFnUpdateStatus and no error when CR has been created without finalizer - Add finalizer",
```
Comment (Disper) [inline internal/controller/runtime/fsm/runtime_fsm_create_kubeconfig_test.go:124]: entry says that it should return `sFnUpdateStatus`  while matcher is `MatchNextFnState: BeNil(),`

File: api/v1/runtime_types.go [hunk 3]
```diff
@@ -33,6 +35,21 @@ const (
 	AnnotationGardenerCloudDelConfirmation = "confirmation.gardener.cloud/deletion"
 )
 
+const (
+	LabelKymaInstanceID      = "kyma-project.io/instance-id"
+	LabelKymaRuntimeID       = "kyma-project.io/runtime-id"
+	LabelKymaShootName       = "kyma-project.io/shootName"
+	LabelKymaRegion          = "kyma-project.io/region"
+	LabelKymaName            = "kyma-project.io/kyma-name"
```
Comment (Disper) [inline api/v1/runtime_types.go:43]: KEB had label key `operator.kyma-project.io/kyma-name` (with `operator.`). Was it intended to be different in KIM?

**Comment by Disper** [comment]
Verification update
After some time, GardenerCluster CR transited to Error state (described as:

```
Status:
  Conditions:
    Last Transition Time:  2024-07-18T12:28:14Z
    Message:               Failed to create secret. Error: secrets "kubeconfig-059dbc39-fd2b-4186-b0e5-8a1bc8ede5b8" already exists
    Reason:                ConditionReasonFailedToCreateSecret
    Status:                False
    Type:                  KubeconfigManagement
  State:                   Error
Events:                    <none>
```
While the Runtime CR is described as:

```
Status:
  Conditions:
    Last Transition Time:  2024-07-18T12:09:51Z
    Message:               Runtime processing completed successfully
    Reason:                ConfigurationCompleted
    Status:                True
    Type:                  Provisioned
    Last Transition Time:  2024-07-18T12:10:19Z
    Message:               Gardener Cluster CR is ready.
    Reason:                GardenerClusterCRReady
    Status:                True
    Type:                  KubeconfigReady
  State:                   Ready
Events:
  Type    Reason                    Age   From                Message
  ----    ------                    ----  ----                -------
  Normal  Initialised               38m   runtime-controller  Runtime initialized: kcp-system/kim-md2
  Normal  Pending                   38m   runtime-controller  Shoot is pending: kcp-system/kim-md2
  Normal  Pending                   37m   runtime-controller  Shoot creation in progress: kcp-system/kim-md2
  Normal  ShootCreationCompleted    30m   runtime-controller  Shoot creation completed: kcp-system/kim-md2
  Normal  GardenerClusterCRCreated  29m   runtime-controller  Gardener Cluster CR created, waiting for readiness: kcp-system/kim-md2
  Normal  GardenerClusterCRReady    29m   runtime-controller  Gardener Cluster CR is ready.: kcp-system/kim-md2
  Normal  ConfigurationCompleted    29m   runtime-controller  Runtime processing completed successfully: kcp-system/kim-md2
```

I would expect the following instead:
1. KIM will not try to create Gardener Cluster CR for the second time resulting in an error in Gardener Cluster CR
2. (I'm not sure about this) Even if Gardener Cluster CR will moved from Ready to Error, the status change should be reflected in Runtime CR condition as well (currently it says that it GardneerCluster CR is Ready)

File: api/v1/runtime_types.go (hunk 3 above)
Comment (koala7659) [reply api/v1/runtime_types.go]: Fixed

Comment (koala7659) [reply api/v1/runtime_types.go]: Also you can see https://github.com/kyma-project/infrastructure-manager/blob/main/docs/adr/001-provisioning.md

**Comment by Disper** [comment]
Probably that's the fix https://github.com/koala7659/infrastructure-manager/pull/2/files. However as @koala7659 is on vacation, I will merge this PR and work on a follow-up.

**Review by Disper — APPROVED** [review]

**Pull request merged**